### PR TITLE
Updates to LES

### PR DESCRIPTION
**Pull request by baperry2**
Updates to LES:
- Error checking to ensure LES is not applied for systems that require models that have not yet been implemented
- Dynamic Smagorinsky: algorithm modified so that everything is calculated at cell-centers, then moved to faces when fluxes are calculated. This removes consistency issues for calculations that previously involved both cell-centered and face quantities. Also the dynamic procedure is now applied to the quantity (CS2/Pr) instead of Pr - this would be equivalent except when the coefficients are filtered to smooth the results, filtering CS2 and Pr separately can cause nonphysical values of CS2/Pr which is the term that appears in the equations.
- Constant Smagorinsky: in the get_sfs_stresses routines, which calculate stresses at faces, the cell-centered density was used previously. This is changed so density is interpolated to the faces before being used.

The changes to the constant Smagorinsky model cause the test "mms-3d-5" to fail because the gold standard file is no longer matched. However, the new results should be more correct than the previous gold standard result.

This request also includes diffusion of AUX variables in PeleC (but not LES yet) and depends on the changes to PelePhysics in my recent pull request for that code. I'm not 100% sure that the changes are properly linked here so that should be checked.

### CODE REVIEW

**Comment by marchdf** [comment]
You should be able to update the PelePhysics submodule as a commit here once that has been merged.

File: Source/PeleC.cpp [hunk 1]
```diff
@@ -134,6 +134,7 @@ int           PeleC::les_test_filter_fgr = 2;
 int           PeleC::comp_Cs2 = 0;
 int           PeleC::comp_CI = PeleC::comp_Cs2 + 1;
 int           PeleC::comp_PrT = PeleC::comp_CI + 1;
+int           PeleC::comp_Cs2ovPrT = PeleC::comp_PrT;
```
Comment (marchdf) [inline Source/PeleC.cpp:138]: Are these two (`comp_PrT` and `comp_Cs2ovPrT`) the same? If so, can we remove one?

Comment (marchdf) [reply Source/PeleC.cpp:138]: Oh I see what's going on. I think this is good.

File: Source/PeleC.cpp [hunk 2]
```diff
@@ -1992,11 +2003,17 @@ void
 PeleC::init_les ()
 {
     // Fill with default coefficient values
-    LES_Coeffs.define(grids, dmap, nCompC, 1);
+    LES_Coeffs.define(grids, dmap, nCompC, 1);
```
Comment (marchdf) [inline Source/PeleC.cpp]: Stray space was added here at the end.

File: Source/PeleC_MOL.cpp [hunk 3]
```diff
@@ -303,7 +316,7 @@ PeleC::getMOLSrcTerm(const amrex::MultiFab& S,
         }  // diffuse_vel
 #endif
       }  // loop over dimension
-
+
```
Comment (marchdf) [inline Source/PeleC_MOL.cpp]: More whitespace added

File: Source/PeleC_les.cpp [hunk 4]
```diff
@@ -237,7 +237,7 @@ PeleC::getSmagorinskyLESTerm (amrex::Real time, amrex::Real dt, amrex::MultiFab&
 
       LESTerm[mfi].setVal(0,vbox,0, NUM_STATE);
       LESTerm[mfi].copy(Lterm,vbox,0,vbox,0,NUM_STATE);
-
+
```
Comment (marchdf) [inline Source/PeleC_les.cpp]: More whitespace added

File: Source/PeleC_les.cpp [hunk 5]
```diff
@@ -354,49 +361,25 @@ PeleC::getDynamicSmagorinskyLESTerm (amrex::Real time, amrex::Real dt, amrex::Mu
...
-          pc_compute_tangential_vel_derivs(cbox.loVect(), cbox.hiVect(),
-                                           dbox.loVect(), dbox.hiVect(),
-                                           BL_TO_FORTRAN_ANYD(Qfab),
-                                           BL_TO_FORTRAN_ANYD(tander_ec[d]),
-                                           geom.CellSize(),&d);
-        }
-#endif
-      } // loop over dimension
-
+      // Tangential derivatives now calculated inside of dynamic_smagorinsky_quantities
```
Comment (marchdf) [inline Source/PeleC_les.cpp]: Remove comment

File: Source/PeleC_les.cpp [hunk 6]
```diff
@@ -431,7 +413,10 @@ PeleC::getDynamicSmagorinskyLESTerm (amrex::Real time, amrex::Real dt, amrex::Mu
       test_filter.apply_filter(g3box, alpha, filtered_alpha);
       test_filter.apply_filter(g3box, flux_T, filtered_flux_T);
 
-      // 4. Calculate the dynamic Smagorinsky coefficients
+      // Again, the tangential derivatives are now dealt with inside the coeffs fortran routine
```
Comment (marchdf) [inline Source/PeleC_les.cpp]: Remove comment

File: Source/PeleC_les.cpp [hunk 7]
```diff
@@ -441,67 +426,92 @@ PeleC::getDynamicSmagorinskyLESTerm (amrex::Real time, amrex::Real dt, amrex::Mu
...
-      // FIXME: REMOVE THIS LATER
-      coeff_cc.setVal(0.16*0.16,g3box,comp_Cs2,1);
-      coeff_cc.setVal(0.09,g3box,comp_CI,1);
-      coeff_cc.setVal(0.7,g3box,comp_PrT,1);
 
-      // 5. Filter to smooth the dynamic coefficients
+      // FIXME: REMOVE THIS LATER
+      // coeff_cc.setVal(0.16*0.16,g3box,comp_Cs2,1);
+      // coeff_cc.setVal(0.09,g3box,comp_CI,1);
+      // coeff_cc.setVal(0.7*0.16*0.16,g3box,comp_PrT,1);
```
Comment (marchdf) [inline Source/PeleC_les.cpp]: Remove these now that you fixed things

File: Source/Src_1d/lesterm_1d.f90 [hunk 8]
```diff
@@ -65,8 +65,6 @@ subroutine pc_smagorinsky_sfs_term(lo,  hi,&
     call build(sfs_eos_state)
```
Comment (marchdf) [inline Source/Src_1d/lesterm_1d.f90:65]: Can we actually just remove `lesterm_1d.f90` and `lesterm_2d.f90` and throw an error for any LES run in anything other than 3D? You already have that check for dynamic smag. I think it should go for any LES.

File: Source/Src_3d/lesterm_3d.f90 [hunk 9]
```diff
@@ -18,7 +52,7 @@ module lesterm_module
        pc_dynamic_smagorinsky_coeffs
 
 contains
-
+
```
Comment (marchdf) [inline Source/Src_3d/lesterm_3d.f90:55]: extra whitespace

File: Source/Src_3d/lesterm_3d.f90 [hunk 10]
```diff
@@ -88,17 +122,12 @@ subroutine pc_smagorinsky_sfs_term(lo,  hi,&
     dxinv = 1.d0/deltax
     Cs2 = Cs**2
     call build(sfs_eos_state)
-
+
```
Comment (marchdf) [inline Source/Src_3d/lesterm_3d.f90]: extra whitespace

File: Source/Src_3d/lesterm_3d.f90 [hunk 11]
```diff
@@ -193,12 +224,12 @@ subroutine pc_smagorinsky_sfs_term(lo,  hi,&
              ! SFS turbulent diffusion
              Uface(:) = HALF*(Q(i,j,k,QU:QW) + Q(i,j,k-1,QU:QW))
              fz(i,j,k,UEDEN) = - sigmazx*Uface(1) - sigmazy*Uface(2) - sigmazz*Uface(3)
-
-             ! SFS heat flux
-             sfs_eos_state % massfrac(:) = Q(i,j,k,QFS:QFS+nspecies-1)
-             sfs_eos_state % T           = Q(i,j,k,QTEMP)
-             call eos_cv(sfs_eos_state)
-             fz(i,j,k,UEDEN) = fz(i,j,k,UEDEN) - sfs_eos_state%gam1 * sfs_eos_state%cv * Cs2 / PrT * flux_T(3)
+
```
Comment (marchdf) [inline Source/Src_3d/lesterm_3d.f90]: extra whitespace

File: Source/Src_3d/lesterm_3d.f90 [hunk 12]
```diff
@@ -454,14 +506,11 @@ subroutine pc_dynamic_smagorinsky_sfs_term(lo,  hi,&
     end do
 
   end subroutine pc_dynamic_smagorinsky_sfs_term
-
-
+
```
Comment (marchdf) [inline Source/Src_3d/lesterm_3d.f90]: extra whitespace

File: Source/Src_3d/lesterm_3d.f90 [hunk 13]
```diff
@@ -636,35 +671,29 @@ subroutine pc_dynamic_smagorinsky_coeffs(lo,  hi,&
     i31 = (1-1) * 3 + 3 ! index for entry (3,1)
     i32 = (2-1) * 3 + 3 ! index for entry (3,2)
     i33 = (3-1) * 3 + 3 ! index for entry (3,3)
-
+
```
Comment (marchdf) [inline Source/Src_3d/lesterm_3d.f90]: extra whitespace

File: Source/Src_3d/lesterm_3d.f90 [hunk 14]
```diff
@@ -785,7 +816,8 @@ subroutine get_sfs_stresses_ydir(i, j, k, Q, Qlo, Qhi, ty, tylo, tyhi, alphaij_y
     S(:,:) = HALF * (dUdx(:,:) + transpose(dUdx(:,:)))
     Skk = S(1,1) + S(2,2) + S(3,3)
     Sijmag = sqrt(TWO * sum(S(:,:)**2))
-    mut = Q(i,j,k,QRHO) * deltabar**2 * Sijmag
+    ! S is located at faces, need to get rho at the face for consistency when calculating mut
```
Comment (marchdf) [inline Source/Src_3d/lesterm_3d.f90]: Remove comment, I think it's obvious, same thing below

File: Source/PeleC.cpp (hunk 1 above)
Comment (baperry2) [reply Source/PeleC.cpp:138]: Added a comment for clarity.

File: Source/Src_1d/lesterm_1d.f90 (hunk 8 above)
Comment (baperry2) [reply Source/Src_1d/lesterm_1d.f90:65]: My preference is to leave them in at least for now, as I may want to do some 2D "LES" for the ignition project eventually. Initially that work will be based on 2D simulations, although in the end we will train/test on 3D simulations.

Comment (marchdf) [reply Source/Src_1d/lesterm_1d.f90:65]: sounds good

**Review by marchdf — APPROVED** [review]
Looks good to me!

**Pull request merged**